### PR TITLE
10 improve intersect functions documentation

### PR DESCRIPTION
**Pull request by Nevsden**
adds content (and possibly solves) #10

### CODE REVIEW

**Comment by jbuckmccready** [comment]
Thanks for the contribution! It looks great, just one or two things I noticed I will put in a review.

File: cavalier_contours/src/core/math/circle_circle_intersect.rs [hunk 1]
```diff
@@ -25,7 +25,29 @@ where
...
+/// # use cavalier_contours::core::math::*;
+/// # use cavalier_contours::core::math::CircleCircleIntr::TangentIntersect;
+/// // Two tangent-intersecting circles of radius `1.0` with euclidean distance of `2.0`
+/// let r1 = 1.0;
+/// let c1 = Vector2::new(0.0, 0.0);
+/// let r2 = 1.0;
+/// let c2 = Vector2::new(0.0, 2.0);
+/// if let CircleCircleIntr::TangentIntersect{point:p} = circle_circle_intr(r1, c1, r2, c2) {
+///    assert_eq!(p, Vector2::new(0.0, 1.0));
+/// }
```
Comment (jbuckmccready) [inline cavalier_contours/src/core/math/circle_circle_intersect.rs:49]: This is really clean to read but will not execute any asserts if the intersect result is not the expected case of `CircleCircleIntr::TangentIntersect`, which may be a bit confusing to the reader. I'm not sure which is cleaner for the example but two ideas come to mind:
Add an else body to the if:
```rust
if let CircleCircleIntr::TangentIntersect{point:p} = circle_circle_intr(r1, c1, r2, c2) {
    assert_eq!(p, Vector2::new(0.0, 1.0));
} else {
    unreachable!("expected tangent intersect");
}
```
Or use the `matches!` macro like the following:
```rust
let intr_result = circle_circle_intr(r1, c1, r2, c2);
assert!(matches!(intr_result, CircleCircleIntr::TangentIntersect{ point } if point == Vector2::new(0.0, 1.0));
```

Maybe there's a better third option...

File: cavalier_contours/src/core/math/line_circle_intersect.rs [hunk 2]
```diff
@@ -28,9 +28,25 @@ where
...
+/// # use cavalier_contours::core::math::*;
+/// # use cavalier_contours::core::math::LineCircleIntr::TangentIntersect;
+/// // A line segment tangent-intersecting a circle with one of the line segments end points.
+/// let p0 = Vector2::new(0.0, 0.0);
+/// let p1 = Vector2::new(1.0, 0.0);
+/// let r = 1.0;
+/// let c = Vector2::new(2.0, 0.0);
+/// if let LineCircleIntr::TangentIntersect{t0: t} = line_circle_intr(p0, p1, r, c) {
+///    assert_eq!(t, 1.0);
+/// }
```
Comment (jbuckmccready) [inline cavalier_contours/src/core/math/line_circle_intersect.rs:48]: Same issue of possibly not running any assert as in the circle_circle_intr case.

File: cavalier_contours/src/core/math/line_line_intersect.rs [hunk 3]
```diff
@@ -34,6 +34,47 @@ where
...
+/// # use cavalier_contours::core::math::LineLineIntr::TrueIntersect;
+/// // A line segment tangent-intersecting a circle with one of the line segments end points.
+/// let v1 = Vector2::new(0.0, 0.0);
+/// let v2 = Vector2::new(1.0, 0.0);
+/// let u1 = Vector2::new(0.5, -1.0);
+/// let u2 = Vector2::new(0.5, 1.0);
+/// if let LineLineIntr::TrueIntersect{seg1_t: t1, seg2_t: t2} = line_line_intr(v1, v2, u1, u2) {
+///    assert_eq!(t1, 0.5);
+///    assert_eq!(t2, 0.5);
+/// }
```
Comment (jbuckmccready) [inline cavalier_contours/src/core/math/line_line_intersect.rs:76]: Same issue of possibly not running any assert as in the circle_circle_intr case.

File: cavalier_contours/src/core/math/line_line_intersect.rs [hunk 4]
```diff
@@ -34,6 +34,47 @@ where
...
+///```
 /// Line segments are defined by `v1->v2` and `u1->u2`.
 /// Handles the cases where the lines may be parallel, collinear, or single points.
```
Comment (jbuckmccready) [inline cavalier_contours/src/core/math/line_line_intersect.rs:79]: These lines should be removed or edited and moved above in the description.

File: cavalier_contours/src/core/math/line_line_intersect.rs [hunk 5]
```diff
@@ -34,6 +34,47 @@ where
...
+/// Either of the following cases:
+/// * Line segments are collinear and overlap
+///
+/// # Examples
+///
+/// ```
+/// # use cavalier_contours::core::traits::*;
+/// # use cavalier_contours::core::math::*;
+/// # use cavalier_contours::core::math::LineLineIntr::TrueIntersect;
+/// // A line segment tangent-intersecting a circle with one of the line segments end points.
```
Comment (jbuckmccready) [inline cavalier_contours/src/core/math/line_line_intersect.rs:68]: Copy paste of comment, it should describe the line-line intersect but was copied from the line-circle intersect doc.

**Review by jbuckmccready — COMMENTED** [review]
I added some comments for changes.

Additionally I have been trying to keep comments within the 100 column Rust formatted width (Rust fmt does not apply to comments so I have been manually editing them into 100 column wrap). If you want to edit them to 100 columns that's helpful, or I'll take care of it, thanks. Here are some pictures of what I mean, the 100 column mark is visible as a vertical line:
image

image

image

File: cavalier_contours/src/core/math/circle_circle_intersect.rs (hunk 1 above)
Comment (Nevsden) [reply cavalier_contours/src/core/math/circle_circle_intersect.rs:49]: I, personally, find the former choice easier to read due to the fact that the `if let` clearly shows the matching arm. Also the nested functions are always harder to read. And as lines do not cost a dime, I always go with the solution that is easier to digest.

Comment (Nevsden) [reply cavalier_contours/src/core/math/circle_circle_intersect.rs:49]: I will embed your suggestions this week.

**Comment by jbuckmccready** [comment]
I pushed some changes to the master branch to fix https://github.com/jbuckmccready/cavalier_contours/issues/20 - I had to add an epsilon parameter to the intersect functions, it may be worth adding something in the documentation about how it controls float fuzzy comparing.

**Comment by Nevsden** [comment]
I will try to embed your comments in the next few days. Due to work I had not have any spare time to finish this issue.

**Comment by jbuckmccready** [comment]
@Nevsden I got things up to date with the added epsilon parameters and made some minor edits, thanks for the contribution!

**Pull request merged**